File: wxwidgets/userifc_py/wxwidgets/resources/wxwidgets/HelloWxwidgetsDemo.cpp
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <unistd.h>
#include "wx/wx.h"
#include "wx/xrc/xmlres.h"
#include "wx/app.h"

class HelloWxwidgetsDemo {
public:
    HelloWxwidgetsDemo(const wxString& title = "HelloWxwidgetsDemo");
    //HelloWxwidgetsDemo(void);
    //HelloWxwidgetsDemo(HelloWxwidgetsDemo const &orig);
    //virtual ~HelloWxwidgetsDemo(void);
    
    void on_frame1_closed(wxCloseEvent& event);
    void on_dialog1_closed(wxCloseEvent& event);
    void on_button1_clicked(wxCommandEvent& event);
    void on_entry1_textChanged(wxCommandEvent& event);
    void on_entry1_returnPressed(wxCommandEvent& event);
    
    wxFrame* frame1;
    
private:
    //wxFrame* frame1;
    wxStaticText* label1;
    wxButton* button1;
    wxTextCtrl* textview1;
    wxDialog* dialog1;
    wxTextCtrl* entry1;
};

HelloWxwidgetsDemo::HelloWxwidgetsDemo(const wxString& title) {
    /*frame1 = new wxFrame(NULL, wxID_ANY, title, wxDefaultPosition,
        wxSize(219, 174), wxDEFAULT_FRAME_STYLE|wxTAB_TRAVERSAL);
    label1 = new wxStaticText(frame1, wxID_ANY, "label1");
    button1 = new wxButton(frame1, wxID_ANY, "button1");
    textview1 = new wxTextCtrl(frame1, wxID_ANY, "",
        wxDefaultPosition, wxSize(160, 80), wxTE_MULTILINE|wxTE_WORDWRAP);
    dialog1 = new wxDialog(NULL, wxID_ANY, "dialog1", wxDefaultPosition,
        wxSize(169, 73), wxDEFAULT_DIALOG_STYLE);
    entry1 = new wxTextCtrl(dialog1, wxID_ANY, "entry1", wxDefaultPosition,
        wxDefaultSize, wxTE_PROCESS_ENTER);
    frame1->SetSizer(new wxBoxSizer(wxVERTICAL));
    dialog1->SetSizer(new wxBoxSizer(wxVERTICAL));
    wxSizer *vbox1 = frame1->GetSizer();
    wxSizer *dialog_vbox1 = dialog1->GetSizer();
    vbox1->Add(label1, 0, wxALIGN_CENTER, 0);
    vbox1->Add(button1, 0, wxALIGN_CENTER, 0);
    vbox1->Add(textview1, 0, wxALIGN_CENTER, 0);
    dialog_vbox1->Add(entry1, 0, wxALIGN_CENTER, 0);
    frame1->Layout();
    dialog1->Layout();
    */
    std::string rsrc_path = getenv("RSRC_PATH") ? getenv("RSRC_PATH") : 
		"resources";
    std::string uiform = rsrc_path + "/" + 
        "wxwidgets/helloForm-wxwidgets.xrc";
    
    wxXmlResource::Get()->InitAllHandlers();
    wxXmlResource::Get()->SetFlags(wxXRC_USE_LOCALE | wxXRC_USE_ENVVARS);
    if (!wxXmlResource::Get()->Load(uiform))
        return;
    frame1 = wxXmlResource::Get()->LoadFrame(NULL, "frame1");
    frame1->SetTitle(title);
    label1 = XRCCTRL(*frame1, "label1", wxStaticText);
    button1 = XRCCTRL(*frame1, "button1", wxButton);
    textview1 = XRCCTRL(*frame1, "textview1", wxTextCtrl);
    dialog1 = wxXmlResource::Get()->LoadDialog(NULL, "dialog1");
    entry1 = XRCCTRL(*dialog1, "entry1", wxTextCtrl);
    
    frame1->Bind(wxEVT_CLOSE_WINDOW, &HelloWxwidgetsDemo::on_frame1_closed,
        this);
    dialog1->Bind(wxEVT_CLOSE_WINDOW, &HelloWxwidgetsDemo::on_dialog1_closed,
        this);
    button1->Bind(wxEVT_BUTTON, &HelloWxwidgetsDemo::on_button1_clicked,
        this);
    entry1->Bind(wxEVT_TEXT, &HelloWxwidgetsDemo::on_entry1_textChanged,
        this);
    entry1->Bind(wxEVT_TEXT_ENTER,
        &HelloWxwidgetsDemo::on_entry1_returnPressed, this);
    
    frame1->Show(true);
}

//HelloWxwidgetsDemo::HelloWxwidgetsDemo(const HelloWxwidgetsDemo& orig) {
//    ;
//}

//HelloWxwidgetsDemo::~HelloWxwidgetsDemo(void) {
//    ;
//}

void HelloWxwidgetsDemo::on_frame1_closed(wxCloseEvent& WXUNUSED(event)) {
    wxExit();
}

void HelloWxwidgetsDemo::on_dialog1_closed(wxCloseEvent& WXUNUSED(event)) {
    frame1->Close();
}

void HelloWxwidgetsDemo::on_button1_clicked(wxCommandEvent& WXUNUSED(event)) {
    frame1->Hide();
    dialog1->Show(true);
    entry1->SetFocus();
    entry1->SetValue("");
}

void HelloWxwidgetsDemo::on_entry1_textChanged(wxCommandEvent& WXUNUSED(event)) {
    ;
}

void HelloWxwidgetsDemo::on_entry1_returnPressed(wxCommandEvent& WXUNUSED(event)) {
    textview1->SetValue("Hello, " + entry1->GetValue() + ".");
    dialog1->Hide();
    frame1->Show(true);
}

class App : public wxApp {
public:
    virtual bool OnInit() wxOVERRIDE;
};

wxIMPLEMENT_APP_NO_MAIN(App);
int main(int argc, char** argv) {
    ////wxString wxVer(wxVERSION_STRING); //wxVer.mb_str().data()
    //const char *wxVer = (new wxString(wxVERSION_STRING))->mb_str().data();
    //fprintf(stderr, "(GCC %d.%d) WxGtk %s GUI\n", __GNUC__, 
	//	__GNUC_MINOR__, wxVer);
    
	wxEntryStart(argc, argv);
    wxTheApp->CallOnInit();
    wxTheApp->OnRun();
    //wxTheApp->MainLoop();
      
    exit(EXIT_SUCCESS);
}
//wxIMPLEMENT_APP(App);
wxIMPLEMENT_WX_THEME_SUPPORT;

bool App::OnInit() {
    if (!wxApp::OnInit())
        return false;
    //wxString wxVer(wxVERSION_STRING); //wxVer.mb_str().data()
    const char *wxVer = (new wxString(wxVERSION_STRING))->mb_str().data();
    char pretextBuf[256];
    snprintf(pretextBuf, sizeof(pretextBuf) - 1,
        "(GCC %d.%d) WxGtk %s GUI\n", __GNUC__, 
		__GNUC_MINOR__, wxVer);
    HelloWxwidgetsDemo *gui = new HelloWxwidgetsDemo("HelloWxwidgetsDemo");
    //this->SetTopWindow(gui->frame1);
    gui->textview1->SetValue(pretextBuf);
    
    return true;
}


// c++ HelloWxwidgetsDemo.cpp -o build/main-HelloWxwidgetsDemo `wx-config --cflags --libs core xrc`
// build/main-HelloWxwidgetsDemo

File: wxwidgets/userifc_py/wxwidgets/resources/wxwidgets/hellowxwidgetsdemo.py
#!/usr/bin/env python -t

from __future__ import (absolute_import, division, print_function,
  unicode_literals)

import sys, os, logging, inspect
from future.builtins import (ascii, filter, hex, map, oct, zip, str)

import wx, wx.xrc

_uiform = 'wxwidgets/helloForm-wxwidgets.xrc'

def userifc_version():
  import platform

  return "(Python {0}) WxGtk {1}".format(platform.python_version(),
    wx.version())

class HelloWxwidgetsDemo(object):
  '''Description:: '''

  def __init__(self):
    '''Construct object'''

    self.app = wx.App(False)
    self.widgets = {}
    
    #self.widgets['frame1'] = wx.Frame(None, title='frame1', name='frame1',
    #  style=wx.DEFAULT_FRAME_STYLE|wx.TAB_TRAVERSAL, size=wx.Size(219, 174))
    #self.widgets['label1'] = wx.StaticText(self.widgets['frame1'], wx.ID_ANY,
    #  'label1', style=wx.ALIGN_CENTER)
    #self.widgets['button1'] = wx.Button(self.widgets['frame1'], wx.ID_ANY,
    #  'button1')
    #self.widgets['textview1'] = wx.TextCtrl(self.widgets['frame1'], wx.ID_ANY,
    #  '', style=wx.TE_MULTILINE | wx.TE_WORDWRAP, size=wx.Size(160, 80))
    #self.widgets['dialog1'] = wx.Dialog(None, title='dialog1',
    #  style=wx.DEFAULT_DIALOG_STYLE, size=wx.Size(169, 73))
    #self.widgets['entry1'] = wx.TextCtrl(self.widgets['dialog1'], wx.ID_ANY,
    #  '', style=wx.TE_PROCESS_ENTER)
    #self.widgets['frame1'].SetSizer(wx.BoxSizer(wx.VERTICAL))
    #self.widgets['dialog1'].SetSizer(wx.BoxSizer(wx.VERTICAL))
    #self.widgets['vbox1'] = self.widgets['frame1'].GetSizer()
    #self.widgets['dialog_vbox1'] = self.widgets['dialog1'].GetSizer()
    #for widget in map(self.widgets.get, ['label1', 'button1', 'textview1']):
    #  self.widgets['vbox1'].Add(widget, 0, wx.ALIGN_CENTER, 0)
    #self.widgets['dialog_vbox1'].Add(self.widgets['entry1'], 0,
    #  wx.ALIGN_CENTER, 0)
    #self.widgets['frame1'].Layout()
    #self.widgets['dialog1'].Layout()
    
    rsrc_path = os.environ.get('RSRC_PATH', 'resources')
    #res = wx.xrc.XmlResource(rsrc_path + '/' + _uiform)
    res = wx.xrc.EmptyXmlResource()
    res.LoadFile(rsrc_path + '/' + _uiform)
    self.widgets['frame1'] = res.LoadFrame(None, 'frame1')
    self.widgets['label1'] = wx.xrc.XRCCTRL(self.widgets['frame1'], 'label1')
    self.widgets['button1'] = wx.xrc.XRCCTRL(self.widgets['frame1'], 
      'button1')
    self.widgets['textview1'] = wx.xrc.XRCCTRL(self.widgets['frame1'],
      'textview1')
    self.widgets['dialog1'] = res.LoadDialog(None, 'dialog1')
    self.widgets['entry1'] = wx.xrc.XRCCTRL(self.widgets['dialog1'], 'entry1')
    
    # connect signal callbacks
    self.widgets['frame1'].Bind(wx.EVT_CLOSE, self.on_frame1_closed, 
      self.widgets['frame1'])
    self.widgets['dialog1'].Bind(wx.EVT_CLOSE, self.on_dialog1_closed, 
      self.widgets['dialog1'])
    self.widgets['button1'].Bind(wx.EVT_BUTTON, self.on_button1_clicked, 
      self.widgets['button1'])
    self.widgets['entry1'].Bind(wx.EVT_TEXT, self.on_entry1_textChanged, 
      self.widgets['entry1'])
    self.widgets['entry1'].Bind(wx.EVT_TEXT_ENTER, 
      self.on_entry1_returnPressed, self.widgets['entry1'])
    
    self.app.SetTopWindow(self.widgets['frame1'])
    self.widgets['frame1'].Show(True)

  def on_frame1_closed(self, frame1, callback_data=None):
    '''Callback for frame1 closed'''
    
    wx.Exit()
  
  def on_dialog1_closed(self, dialog1, callback_data=None):
    '''Callback for dialog1 closed'''
    
    self.widgets['frame1'].Close()
  
  def on_button1_clicked(self, button1, callback_data=None):
    '''Callback for button1 clicked'''
    
    self.widgets['frame1'].Hide()
    self.widgets['dialog1'].Show(True)
    self.widgets['entry1'].SetFocus()
    self.widgets['entry1'].SetValue('')
  
  def on_entry1_textChanged(self, entry1, callback_data=None):
    '''Callback for entry1 textChanged'''
    
    
  
  def on_entry1_returnPressed(self, entry1, callback_data=None):
    '''Callback for entry1 returnPressed'''
    
    self.widgets['textview1'].SetValue("Hello, " + 
      self.widgets['entry1'].Value + ".")
    self.widgets['dialog1'].Hide()
    self.widgets['frame1'].Show(True)

def main(argv=None):
  pretext = '{0} GUI\n'.format(userifc_version())
  gui = HelloWxwidgetsDemo()
  gui.widgets['textview1'].SetValue(pretext)
  
  gui.app.MainLoop()
  return 0

if '__main__' == __name__:
  sys.exit(main(sys.argv[1:]))


# (one-time): pip install --user wxpython
# python hellowxwidgetsdemo.py
